### PR TITLE
Add tables and changes for plate map tracking

### PR DESCRIPTION
**Pull request by squirrelo**
This adds the database infrastructure for creating and storing plate map and wet lab workflow information.

### CODE REVIEW

**Comment by squirrelo** [comment]
@wasade @josenavas could you give this a lookover?

**Comment by josenavas** [comment]
This PR is missing stuff - better if we can discuss this in person later.

**Comment by squirrelo** [comment]
What is it missing?

File: amgut/db/patches/0029.sql [hunk 1]
```diff
@@ -0,0 +1,121 @@
+-- October 12, 2015
```
Comment (josenavas) [inline amgut/db/patches/0029.sql]: Just a general comment - is this dropping information w/o moving it to the new structure? Doesn't it looks dangerous?

Comment (squirrelo) [inline amgut/db/patches/0029.sql]: Yes, it's dropping plate information, but that information is garbled and not up to date. This table was never used correctly in the old database.

File: amgut/db/patches/0029.sql [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+CREATE TABLE barcodes.robots ( 
+	robot_id             bigserial  NOT NULL,
+	robot_type           varchar  NOT NULL,
+	robot_name           varchar(100)  NOT NULL,
+	date_added           date DEFAULT current_date NOT NULL,
+	CONSTRAINT pk_robots PRIMARY KEY ( robot_id )
+ ) ;
+
+CREATE TABLE barcodes.pcr_plate (
```
Comment (josenavas) [inline amgut/db/patches/0029.sql]: I discourage the usage of quoted columns names, it can add complexity on the python code...

Comment (squirrelo) [inline amgut/db/patches/0029.sql]: ah, that was a missed underscore. Will fix.

Comment (josenavas) [inline amgut/db/patches/0029.sql]: It also applies for the numbers - you can probably re-order the column name: tool_300_8

Comment (squirrelo) [inline amgut/db/patches/0029.sql]: already done :)

**Comment by squirrelo** [comment]
Added tables/cols from conversation, should be ready to go.

**Comment by squirrelo** [comment]
Realized this is missing a table, one moment while adding it

File: amgut/db/patches/0029.sql [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+	lot_id               bigserial  NOT NULL,
+	lot_type             varchar  NOT NULL,
+	lot                  varchar  NOT NULL,
+	date_added           date DEFAULT current_date NOT NULL,
+	CONSTRAINT pk_lots PRIMARY KEY ( lot_id )
+ );
+
+CREATE TABLE barcodes.robots ( 
+	robot_id             bigserial  NOT NULL,
+	robot_type           varchar  NOT NULL,
```
Comment (wasade) [inline amgut/db/patches/0029.sql]: should this and `lot_type` be look ups?

File: amgut/db/patches/0029.sql [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+	robot_name           varchar(100)  NOT NULL,
```
Comment (wasade) [inline amgut/db/patches/0029.sql]: this is effectively the ID; is there a reason to not use it as it's a natural pk?

File: amgut/db/patches/0029.sql [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+	person_id            bigint  NOT NULL,
+	name                 varchar(100)  NOT NULL,
+	email                varchar  NOT NULL,
+	date_added           date DEFAULT current_date NOT NULL,
+	CONSTRAINT pk_person PRIMARY KEY ( person_id )
+ );
+
+CREATE TABLE barcodes.protocol ( 
+	protocol_id          bigserial  NOT NULL,
+	protocol_name        varchar(100)  NOT NULL,
```
Comment (wasade) [inline amgut/db/patches/0029.sql]: same here regarding pk

File: amgut/db/patches/0029.sql [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+	template_dna         integer  NOT NULL,
+	platform             varchar  NOT NULL,
+	pcr_primers          varchar  NOT NULL,
+	instrument_model     varchar  NOT NULL,
+	linker               varchar  NOT NULL,
+	sequencing_method    varchar  NOT NULL,
+	target_gene          varchar  NOT NULL,
```
Comment (wasade) [inline amgut/db/patches/0029.sql]: what happens for wgs, etc?

File: amgut/db/patches/0029.sql [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+	CONSTRAINT pk_person PRIMARY KEY ( person_id )
+ );
+
+CREATE TABLE barcodes.protocol ( 
+	protocol_id          bigserial  NOT NULL,
+	protocol_name        varchar(100)  NOT NULL,
+	template_dna         integer  NOT NULL,
+	platform             varchar  NOT NULL,
+	pcr_primers          varchar  NOT NULL,
+	instrument_model     varchar  NOT NULL,
```
Comment (wasade) [inline amgut/db/patches/0029.sql]: presumably a lookup?

File: amgut/db/patches/0029.sql [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+	target_gene          varchar  NOT NULL,
+	target_subfragment   varchar  NOT NULL,
+	CONSTRAINT pk_protocol PRIMARY KEY ( protocol_id )
+ );
+COMMENT ON COLUMN barcodes.protocol.template_dna IS 'Template DNA added, in uL';
+
+CREATE TABLE barcodes.pcr_plate_run ( 
+	run_id               bigint  NOT NULL,
+	barcode              varchar  NOT NULL,
+	nickname             varchar(100)  NOT NULL,
```
Comment (wasade) [inline amgut/db/patches/0029.sql]: ?

File: amgut/db/patches/0029.sql (hunk 7 above)
Comment (squirrelo) [inline amgut/db/patches/0029.sql]: enum, but yeah

File: amgut/db/patches/0029.sql (hunk 8 above)
Comment (squirrelo) [inline amgut/db/patches/0029.sql]: multiple pcr plates from a single extraction plate, so need a way to track them. They are not kept, so they don't get a barcode, but we still need to differentiate them.

File: amgut/db/patches/0029.sql (hunk 3 above)
Comment (squirrelo) [inline amgut/db/patches/0029.sql]: They will be ENUMS, but not adding that info until we settle on a design first.

File: amgut/db/patches/0029.sql (hunk 4 above)
Comment (squirrelo) [inline amgut/db/patches/0029.sql]: True, will do that.

File: amgut/db/patches/0029.sql (hunk 6 above)
Comment (squirrelo) [inline amgut/db/patches/0029.sql]: `target_gene = ALL`?

**Comment by squirrelo** [comment]
Ready for another round of review. The two HTML pages have been broken up by schema, since the barcodes schema started to get really large.

**Comment by squirrelo** [comment]
Another question: do we want to add the ENUMs on this pull request or on another pull request where we add the initial data?

**Comment by wasade** [comment]
I think it conceptually fits better in here

**Comment by squirrelo** [comment]
Ok, will add here then.

**Comment by wasade** [comment]
Thanks!

**Comment by squirrelo** [comment]
enums now in, should be good for a full review/merge if OK.

File: amgut/db/patches/0029.sql (hunk 8 above)
Comment (squirrelo) [inline amgut/db/patches/0029.sql]: Oh, misread where this was. Yes, this is duplicate of center_project_name now.

Comment (squirrelo) [inline amgut/db/patches/0029.sql]: Sorry, first comment stands. This is needed to track the PCR plate.

**Comment by squirrelo** [comment]
Anyone have time to review this? @josenavas @wasade @mortonjt

**Comment by squirrelo** [comment]
Need to make more changes in light of new info

**Comment by squirrelo** [comment]
This should now address the issues that came up and be ready to go.

**Comment by squirrelo** [comment]
Closed for redesign.

**Pull request closed without merging**